deployer/deployer-services/src/main/java/io/hyscale/deployer/services/handler/impl/V1PodHandler.java

```diff
@@ -372,4 +385,146 @@ public boolean cleanUp() {
+			latestPodSelector = v1DeploymentHandler.getPodSelector(apiClient, selector, true, namespace);
+			replicas = v1Deployment != null ? v1Deployment.getSpec().getReplicas() : 1;
+		}
+		Watch<V1Pod> watch = null;
+		String namespaceSelector=ResourceSelectorUtil.getNamespaceSelector(namespace); 
+
+		Set<String> containersReadyPods = new HashSet<String>();
+		Set<String> readyPods = new HashSet<String>();
+		Set<String> initializedPods = new HashSet<String>();
+		Set<String> createdPods = new HashSet<String>();
+		
+		ActivityContext initializedActivityContext = null;
+		ActivityContext creationActivityContext = null;
+		ActivityContext readyActivityContext = null;
+
+
+		watch = Watch.createWatch(apiClient,
+				api.listPodForAllNamespacesCall(null, null,namespaceSelector , latestPodSelector, null, null, null, replicas*10, Boolean.TRUE, null),
+				new TypeToken<Watch.Response<V1Pod>>() {
+				}.getType());
+		try {
+			initializedActivityContext = new ActivityContext(DeployerActivity.POD_INITIALIZED);
+			WorkflowLogger.startActivity(initializedActivityContext);
+
+			for (Watch.Response<V1Pod> item : watch) {
+				if (item.object.getStatus().getContainerStatuses().stream()
+						.anyMatch(s -> s.getRestartCount() > POD_RESTART_COUNT)) {
+					WorkflowLogger.endActivity(Status.FAILED);
+					throw new HyscaleException(DeployerErrorCodes.FAILED_TO_INITIALIZE_POD);
+				}
+
+				if (PodPredicates.isPodFailed().test(item.object)) {
+					WorkflowLogger.endActivity(Status.FAILED);
+					throw new HyscaleException(DeployerErrorCodes.FAILED_TO_INITIALIZE_POD);
+				}
+				
+				if (PodPredicates.isPodInitialized().test(item.object)) {
+					initializedPods.add(item.object.getMetadata().getName());
+				}
+				
+				if (PodPredicates.isPodCreated().test(item.object)) {
+					createdPods.add(item.object.getMetadata().getName());
+				}
+				
+				if (PodPredicates.isPodReady().test(item.object)) {
+					readyPods.add(item.object.getMetadata().getName());
+				}
+
+				if (containersReadyPods.size() == replicas) {
+					break;
+				}
+			}
+			
+			if (initializedPods.size() == replicas) {
+				WorkflowLogger.endActivity(initializedActivityContext, Status.DONE);
+			}else {
+				throw new HyscaleException(DeployerErrorCodes.FAILED_TO_INITIALIZE_POD);
+			}
+			
+			creationActivityContext = new ActivityContext(DeployerActivity.POD_CREATION);
+			WorkflowLogger.startActivity(creationActivityContext);
+			if(createdPods.size()==replicas) {
+				WorkflowLogger.endActivity(creationActivityContext, Status.DONE);
+			}else {
+				throw new HyscaleException(DeployerErrorCodes.FAILED_TO_CREATE_POD);
+			}
+			
+			readyActivityContext = new ActivityContext(DeployerActivity.POD_READINESS);
+			WorkflowLogger.startActivity(readyActivityContext);
+			if(readyPods.size()==replicas) {
+				WorkflowLogger.endActivity(readyActivityContext, Status.DONE);
+			}else {
+				throw new HyscaleException(DeployerErrorCodes.POD_FAILED_READINESS);
+			}
```

Comment:
All these activities should happen along with the watch events 
In this logic it being ended after all events are processed ...it is not live with the events